### PR TITLE
GPJax and flexible train/valid losses

### LINKED ISSUE #2: Allow for flexible training and validation losses
It would be nice to allow the user to customize the training and validation losses (e.g. to include regularization terms). This shouldn't be a difficult enhancement.

### PR DESCRIPTION
**Pull request by jgbarcos**
## Added GPJax and allow train and valid losses to be passed as arguments to ift

- Resolves #2 
- Provides initial implementation for #1.
- Fixes the inverse transform missing the epsilon
- Adds test for GPJax similar to the existing GP tests
- Updates the notebook to use both the original gp and the gpjax.

## Comparison
I've run a manual comparison on the notebook by fixing the seed and running the first problem on master and this PR.

Seems that using GPJax deteriorates in performance, either because of the existing GP implementation being so barebones that is faster and/or simply I'm missing some jax jit or performance trick. Qualitatively, pr-gpjax meta-parameter results are close enough with respect to master-gp. Using the flexible train/valid losses doesn't change anything (master vs pr-gp) which was expected.

### Performance
master-gp: 2.15it/s
pr-gp:     2.15it/s
pr-gpjax:  1.48it/s

### Final meta-params
master-gp: {'c_exp': 0.30767794, 'c_log': -0.11056437, 'c_sqrt': 0.24794468, 'c_x': 0.18069236}
pr-gp:     {'c_exp': 0.30767794, 'c_log': -0.11056437, 'c_sqrt': 0.24794468, 'c_x': 0.18069236}
pr-gpjax:  {'c_exp': 0.30767789, 'c_log': -0.11056453, 'c_sqrt': 0.2479447, 'c_x': 0.18069223}

## Discussion
Regarding my concerns about how to customize the GP (e.g. fixed non-learnable params or different kernels), now that the losses become user-defined and passed as arguments, the ift is now agnostic to the GP model completely. It is possible to do plenty of things like: adapt/meta partitions, set non-learnable params or even use a different kernel/model; by a clever use of defining the initial parameters, a combine_func and the losses (inside you can define your own gp_module functional calls).  I think the notebook already covers these cases implicitly.

So right now, providing the gp.py or gpjax.py shouldn't even be needed other than as an example, as in practice you will probably define your own GP (e.g. using a different kernel) to suit your problem.

Another approach is to focus a bit more on GPJax and facilitate its use, as we can take advantage of its initialize method and use the parameter information that is automatically generated (such as learnable params and parameter transformations).

### CODE REVIEW

**Comment by jgbarcos** [comment]
Forgot to put the gpjax dependency. Also I found that there is newer gpjax version (this PR was for 0.5.9), I could try to bump up if needed.

File: adkf/gp.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ def standard_positive_transform(x: jnp.array) -> jnp.array:
 
 def standard_inverse_transform(x: jnp.array) -> jnp.array:
     """Implements inverse softplus transform."""
-    return jnp.log(jnp.exp(x) - 1)
+    return jnp.log(jnp.exp(x - 1e-6) - 1)
```
Comment (AustinT) [inline adkf/gp.py:18]: Maybe add a comment saying this is to under the addition of 1e-6 in the forward transform? Better yet, maybe define `EPS = 1e-6` and use it in both places?

File: adkf/gpjax.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+def setup_gp(num_datapoints: int):
+    """Define the GP model."""
+    prior = gpx.Prior(kernel=jk.RBF())
+    likelihood = gpx.Gaussian(num_datapoints=num_datapoints)
+    return prior, likelihood
+
+
+def ift_to_gpx(ift_params: Any) -> Any:
+    """Applies"""
```
Comment (AustinT) [inline adkf/gpjax.py:30]: Did you forget to finish this comment?

File: adkf/ift.py [hunk 3]
```diff
@@ -3,56 +3,39 @@
...
+        Any,
+        param_combine_type,
+        jnp.ndarray,
+        jnp.ndarray,
+        jnp.ndarray,
+        jnp.ndarray,
+        nn.Module,
+    ],
+    float,
+]
```
Comment (AustinT) [inline adkf/ift.py:38]: Thanks for generalizing this, it looks good!

File: adkf/ift.py [hunk 4]
```diff
@@ -62,6 +45,7 @@ def optimize_train_loss_adam(
     x_train: jnp.ndarray,
     y_train: jnp.ndarray,
     feature_extractor: nn.Module,
+    adapt_loss: adapt_loss_type,
```
Comment (AustinT) [inline adkf/ift.py:48]: Do you think it would make sense to have the train mll as a default here?

File: adkf/ift.py [hunk 5]
```diff
@@ -169,13 +156,22 @@ def ift_gradient_update(
...
     hes_array, _ = jax.flatten_util.ravel_pytree(hes)
     cross_array, _ = jax.flatten_util.ravel_pytree(cross_derivs)
+
+    hes_matrix = hes_array.reshape(len(adapt_params_flat), len(adapt_params_flat))
+    cross_matrix = cross_array.reshape(len(adapt_params_flat), len(meta_params_flat))
+
+    if fix_singular_hessian and jnp.linalg.det(hes_matrix) == 0:
+        eig_val, eig_vec = jnp.linalg.eigh(hes_matrix)
+        new_eig_val = eig_val.at[eig_val == 0].set(1e-6)
+        hes_matrix = eig_vec @ jnp.diag(new_eig_val) @ jnp.linalg.inv(eig_vec)
```
Comment (AustinT) [inline adkf/ift.py:166]: Isn't this a very expensive check? Maybe this is why the iteration speed is slower. I would personally not check for this. Did you find that it was necessary?

File: adkf/ift.py [hunk 6]
```diff
@@ -169,13 +156,22 @@ def ift_gradient_update(
...
     d_adapt_by_d_meta = -jnp.linalg.solve(
-        hes_array.reshape(len(adapt_params_flat), len(adapt_params_flat)),
-        cross_array.reshape(len(adapt_params_flat), len(meta_params_flat)),
+        hes_matrix,
+        cross_matrix,
     )
 
     # Find gradients of validation loss
-    L_V = pred_mll_loss  # could change this to be a parameter
+    L_V = meta_loss
```
Comment (AustinT) [inline adkf/ift.py:174]: Maybe just refactor out `L_V`?

File: adkf/ift.py [hunk 7]
```diff
@@ -148,7 +135,7 @@ def ift_gradient_update(
     meta_params_flat, meta_unflatten = jax.flatten_util.ravel_pytree(meta_params)
 
     # Form Hessian and cross-derivatives
-    L_T = train_mll_loss  # could change this to be a parameter
+    L_T = adapt_loss
```
Comment (AustinT) [inline adkf/ift.py:138]: Maybe just refactor out `L_T`?

File: environment.yml [hunk 8]
```diff
@@ -19,3 +22,4 @@ dependencies:
     - flax
     - optax
     - jaxopt
+    - gpjax==0.5.9
```
Comment (AustinT) [inline environment.yml:25]: Why this specific version? Maybe put a comment.

File: environment.yml [hunk 9]
```diff
@@ -1,14 +1,17 @@
...
   - conda-forge
+  - nvidia
 dependencies:
   - python=3.8
   - pip
   # Standard data science stuff
   - numpy
   - tqdm
   # ML libraries
+  - jaxlib=*=*cuda*
```
Comment (AustinT) [inline environment.yml:12]: Why did you put `*cuda*`? What happens if you omit it? What about a CPU-only install?

File: tests/test_gpjax.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    x, y = xy_train
+    mll_output = gpjax.train_mll(x, y, gp_params)
+    assert jnp.isclose(mll_output, -3.59153)
+
+
+def test_pred_mll(xy_train, xy_test, gp_params):
+    x, y = xy_train
+    x_q, y_q = xy_test
+    pred_mll = gpjax.predictive_mll(x_q, y_q, x, y, gp_params)
+    assert jnp.isclose(pred_mll, -1.3878)
```
Comment (AustinT) [inline tests/test_gpjax.py:42]: Maybe add a comment explaining that these are copied from the other tests? If we polish this repo more it would make sense to merge them.

**Review by AustinT — APPROVED** [review]
This looks really good, thanks Javier! I left a few comments, mostly minor things (e.g. adding some comments or explanation). My only real concern is checking the determinant of the Hessian at every step, which I guess is probably the cause of the slower performance.

Overall I think you are correct that the IFT code now functions in a standalone way and is agnostic to the actual GP implementation. Maybe that could be released by itself?
